File: server/modules/git/exec.ts
import { execFileSync } from "node:child_process";

export class GitError extends Error {
  constructor(
    message: string,
    readonly stderr: string,
  ) {
    super(message);
    this.name = "GitError";
  }
}

function readGitOutput(
  chunk: string | Buffer | undefined,
  fallback: string,
): string {
  if (typeof chunk === "string") return chunk;
  return chunk?.toString("utf-8") ?? fallback;
}

export function runGit(
  repoPath: string,
  args: string[],
  options?: { trim?: boolean },
): string {
  try {
    const out = execFileSync("git", ["-C", repoPath, ...args], {
      encoding: "utf-8",
      stdio: ["ignore", "pipe", "pipe"],
    });
    return options?.trim === false ? out : out.trim();
  } catch (err) {
    const e = err as {
      status?: number;
      stdout?: string | Buffer;
      stderr?: string | Buffer;
      message?: string;
    };
    const stderr = readGitOutput(e.stderr, e.message ?? "git failed").trim();
    throw new GitError(stderr || "git command failed", stderr);
  }
}

/** Like runGit, but treats exit code 1 with stdout as success (e.g. `git diff`). */
export function runGitWithStdout(
  repoPath: string,
  args: string[],
  options?: { trim?: boolean },
): string {
  try {
    const out = execFileSync("git", ["-C", repoPath, ...args], {
      encoding: "utf-8",
      stdio: ["ignore", "pipe", "pipe"],
    });
    return options?.trim === false ? out : out.trim();
  } catch (err) {
    const e = err as {
      status?: number;
      stdout?: string | Buffer;
      stderr?: string | Buffer;
      message?: string;
    };
    const stdout = readGitOutput(e.stdout, "");
    if (e.status === 1 && stdout) {
      return options?.trim === false ? stdout : stdout.trim();
    }
    const stderr = readGitOutput(e.stderr, e.message ?? "git failed").trim();
    throw new GitError(stderr || "git command failed", stderr);
  }
}

export function isGitRepo(repoPath: string): boolean {
  try {
    runGit(repoPath, ["rev-parse", "--git-dir"]);
    return true;
  } catch {
    return false;
  }
}

File: server-rs/src/git/exec.rs
use std::process::Command;

#[derive(Debug, thiserror::Error)]
#[error("{message}")]
pub struct GitError {
    pub message: String,
    pub stderr: String,
}

impl GitError {
    pub fn new(message: impl Into<String>, stderr: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            stderr: stderr.into(),
        }
    }
}

pub struct RunGitOptions {
    pub trim: bool,
}

impl Default for RunGitOptions {
    fn default() -> Self {
        Self { trim: true }
    }
}

pub fn run_git(repo_path: &str, args: &[&str], options: Option<RunGitOptions>) -> Result<String, GitError> {
    let opts = options.unwrap_or_default();
    let output = Command::new("git")
        .arg("-C")
        .arg(repo_path)
        .args(args)
        .output()
        .map_err(|e| GitError::new(e.to_string(), ""))?;

    if output.status.success() {
        let out = String::from_utf8_lossy(&output.stdout).into_owned();
        return Ok(if opts.trim { out.trim().to_string() } else { out });
    }

    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
    Err(GitError::new(
        if stderr.is_empty() {
            "git command failed".to_string()
        } else {
            stderr.clone()
        },
        stderr,
    ))
}

pub fn run_git_with_stdout(
    repo_path: &str,
    args: &[&str],
    options: Option<RunGitOptions>,
) -> Result<String, GitError> {
    let opts = options.unwrap_or_default();
    let output = Command::new("git")
        .arg("-C")
        .arg(repo_path)
        .args(args)
        .output()
        .map_err(|e| GitError::new(e.to_string(), ""))?;

    if output.status.success() {
        let out = String::from_utf8_lossy(&output.stdout).into_owned();
        return Ok(if opts.trim { out.trim().to_string() } else { out });
    }

    let stdout = String::from_utf8_lossy(&output.stdout).into_owned();
    if output.status.code() == Some(1) && !stdout.is_empty() {
        return Ok(if opts.trim {
            stdout.trim().to_string()
        } else {
            stdout
        });
    }

    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
    Err(GitError::new(
        if stderr.is_empty() {
            "git command failed".to_string()
        } else {
            stderr.clone()
        },
        stderr,
    ))
}

pub fn is_git_repo(repo_path: &str) -> bool {
    run_git(repo_path, &["rev-parse", "--git-dir"], None).is_ok()
}
